lib/tcx.js

```diff
@@ -0,0 +1,207 @@
+import { nodeVal, get1 } from "./shared";
+
+const attributeNames = [
+  // Trackpoint attributes
+  ["heartRate", "heartRates"],
+  ["cadence", "cadences"],
+  // Extended Trackpoint attributes
+  ["Speed", "speeds"],
+  ["Watts", "watts"],
+  // Lap attributes
+  ["TotalTimeSeconds", "totalTimeSeconds"],
+  ["DistanceMeters", "distanceMeters"],
+  ["MaximumSpeed", "maxSpeed"],
+  ["AverageHeartRateBpm", "avgHeartRate"],
+  ["MaximumHeartRateBpm", "maxHeartRate"],
+  // Extended Lap attributes
+  ["AvgSpeed", "avgSpeed"],
+  ["AvgWatts", "avgWatts"],
+  ["MaxWatts", "maxWatts"],
+];
+
+function getProperties(node) {
+  const prop = {};
+  const extensions = node.getElementsByTagNameNS(
+    "http://www.garmin.com/xmlschemas/ActivityExtension/v2",
+    "*"
+  );
+  attributeNames.forEach((attr) => {
+    const raw = get1(node, attr[0]);
+    if (raw !== null) {
+      prop[attr[1]] = parseFloat(nodeVal(raw));
+    } else if (extensions !== null) {
+      for (let i = 0; i < extensions.length; i++) {
+        const extension = extensions[i];
+        if (extension.localName === attr[0]) {
+          const ext = nodeVal(extension);
+          if (ext !== null) {
+            const v = parseFloat(ext);
+            if (!isNaN(v)) {
+              prop[attr[1]] = v;
+            }
+          }
+        }
+      }
+    }
+  });
+  return prop;
+}
+
+function coordPair(x) {
+  const lon = nodeVal(get1(x, "LongitudeDegrees"));
+  const lat = nodeVal(get1(x, "LatitudeDegrees"));
+  if (!lon.length || !lat.length) {
+    return null;
+  }
+  const ll = [parseFloat(lon), parseFloat(lat)];
+  const alt = get1(x, "AltitudeMeters");
+  const heartRate = get1(x, "HeartRateBpm");
+  const cadence = get1(x, "Cadence");
+  const time = get1(x, "Time");
+  let a;
+  if (alt) {
+    a = parseFloat(nodeVal(alt));
+    if (!isNaN(a)) {
+      ll.push(a);
+    }
+  }
+  const result = {
+    coordinates: ll,
+    time: time ? nodeVal(time) : null,
+    heartRate: heartRate ? parseFloat(nodeVal(heartRate)) : null,
+    cadence: cadence ? parseFloat(nodeVal(cadence)) : null,
+  };
+  const extensions = x.getElementsByTagNameNS(
+    "http://www.garmin.com/xmlschemas/ActivityExtension/v2",
+    "*"
+  );
+  if (extensions !== null) {
+    for (let i = 0; i < extensions.length; i++) {
+      const extension = extensions[i];
+      attributeNames
+        .map((r) => r[0])
+        .filter((n) => n === extension.localName)
```

Comment:
This seems like what is desired is not to get all namespaced tags and filter them by attributeNames, but rather to call `getElementByTagNameNS` with each attributeName and the namespace.